core/src/main/java/org/apache/iceberg/util/SnapshotUtil.java

```diff
@@ -65,12 +65,16 @@ public static boolean ancestorOf(Table table, long snapshotId, long ancestorSnap
   }
 
   /**
-   * Traverses the history of the table's current snapshot and finds the oldest Snapshot.
-   * @return null if there is no current snapshot in the table, else the oldest Snapshot.
+   * Traverses the history of the table's current snapshot and finds the oldest Snapshot after the timestamp.
+   * @return null if there is no current snapshot in the table, else the oldest Snapshot after the timestamp.
```

Comment:
Thanks for the comment we will add the milliseconds in the documentation as suggested
oldestSnapshot will grab the snapshot that is oldest which just crosses the timestamp.
Let's take an example with the assumption
Snapshot with time series: 10,20,30,40,50
Current snapshot will be with timestamp 50
The input timestamp is 31.
So the return snapshot will return with timestamp 40

Reply:
Ok. That does help clarify for me @zhengzengprc

I think I was expecting it to work slightly differently. I'd like to clarify what is the exact intent / use case for choosing to grab the snapshot that is strictly greater than the passed in timestamp?

And for your present use case, what is supposed to happen if a user passes in a timestamp that exists exactly?

let's say we have snapshot with time series: 10,20,30,40,50
Current snapshot will be with timestamp 50.
The input timestamp is 30.
Will the oldestSnapshot still return with timestamp 40, or will the user get timestamp 30?

For your proposed functionality of grabbing the timestamp just after that passed in, I'm guessing it would return 4 still (as that's the newest snapshot just after 31)

These are my two biggest outstanding questions / wondering why the motivation for the feature the way that it is.

Some of the engines (or batch for spark) support `as-of-timestamp`, which I believe should definitely grab 30 in the case I mentioned as that's the exact timestamp. I believe it would _also_ grab 30 in your example of passing in 31 as the input timestamp, as the table's state as of 31 is what it was at timestamp 30.

**Question**: What's the use case / motivation for grabbing the timestamp just _after_ the one passed in, and not grabbing the table as of the timestamp that is given (or starting from the snapshot that was the most recent as of the timestamp - so grabbing 30 in your example of input 31 instead of grabbing 40)? What problem does this solve that requires it to differ from the semantics of `as-of-timestamp`? And could the problem that you're looking to solve be solved by `as-of-timestamp`?

Reply:
Hello @kbendick , thanks for bringing up the questions. That's a really good question for the strictly greater than the passed in timestamp.
Honestly, I don't have a strong reason use strictly greater. IMO, the function should include the "equal" case while user passes in the existing timestamp that exists.
From the sample:
let's say we have snapshot with time series: 10,20,30,40,50
Current snapshot will be with timestamp 50.
The input timestamp is 30.
I think 30 should be the answer from the function calls.
The boolean logic should include the "=" case,
`while (current.parentId() != null && table.snapshot(current.parentId()).timestampMillis() >= timestamp)`
I try to reach @SreeramGarlapati to see if we have a strong reason put "strictly greater", if I miss something here. But I am still waiting input from Sreeram.
@SreeramGarlapati do you know anything behind the scene that we put the "strictly greater"?
Personally I prefer the ">=" approach.
Beside the reason of `as-of-timestamp`. I think the function return should be consistent, let's take back above example to go through current implementation:
Case 1:
Snapshot with time series: 10,20,30,40,50
Current timestamp: 50
Input timestamp: 50
Function return 50

Case 2:
Snapshot with time series: 10,20,30,40,50
Current timestamp: 50
Input timestamp: 40
Function return 50

The result looks like inconsistent to me, since case 1 returns inclusive 50, but case 2 not. IMO, these logic should be consistent with each other

Reply:
Confirm from Sreeram that we should NOT use strictly greater than in the logic, so the next iteration include the equal case. Thanks for the comment @kbendick

Reply:
Hi @zhengzengprc sorry for the delay in reply. I was out sick part of last week.

I'll read through this tonight and get back to you tomorrow. I think `as-of-timestamp` actually has the strictly greater than logic (same as this) which I am going to test out.